### DISCUSSION TITLE
'--pdf-engine', '-t pdf' and '-t html'

### DISCUSSION THREAD

**Discussion by johnmapeson** [jgm/pandoc#10107] [Q&A] [answered]
> `--pdf-engine=PROGRAM`
>
> Use the specified engine when producing PDF output. Valid values are `pdflatex`, `lualatex`, `xelatex`, `latexmk`, `tectonic`, `wkhtmltopdf`, `weasyprint`, `pagedjs-cli`, `prince`, `context`, `pdfroff`, and `typst`. If the engine is not in your PATH, the full path of the engine may be specified here. If this option is not specified, pandoc uses the following defaults depending on the output format specified using [`-t/--to`](https://pandoc.org/MANUAL.html#option--to):
>
> - [`-t latex`](https://pandoc.org/MANUAL.html#option--to) or none: `pdflatex` (other options: `xelatex`, `lualatex`, `tectonic`, `latexmk`)
> - [`-t context`](https://pandoc.org/MANUAL.html#option--to): `context`
> - [`-t html`](https://pandoc.org/MANUAL.html#option--to): `wkhtmltopdf` (other options: `prince`, `weasyprint`, `pagedjs-cli`; see [print-css.rocks](https://print-css.rocks/) for a good introduction to PDF generation from HTML/CSS)
> - [`-t ms`](https://pandoc.org/MANUAL.html#option--to): `pdfroff`
> - [`-t typst`](https://pandoc.org/MANUAL.html#option--to): `typst`

I'm very confused.

The first question. Isn't it the most appropriate use of `--pdf-engine` is when we are going to produce a PDF file, e.g.

```
pandoc -o output.pdf --pdf-engine=prince input.htm
```

And if yes, then why `-t pdf` is not mentioned in the list above?

The second question. Any for what purpose `--pdf-engine` in the list above is listed for e.g. producing HTML file? As far as I understand, it isn't required to use a pdf engine to produce an HTMl file, isn't it?

**Comment by badumont** [comment, marked as answer]
Pandoc doesn't produce a PDF directly: it first generates an intermediary file, which it post-processes with the program set by `--pdf-engine`. When the output file has the `.pdf` extension, the `-t` option specifies the format of the intermediary file. So a full command would be e.g.:

  pandoc -o myfile.pdf -t html --pdf-engine=prince myfile.md

Now, you can omit some parts of the command, and Pandoc will guess how to fill the blanks. For instance, the manual says that if you write something like `pandoc -o myfile.pdf -t context myfile.md`, without setting `--pdf-engine`, Pandoc will guess that the value of `--pdf-engine` is `context`. If you omit `--pdf-engine=prince` in the example above, it will fall back to `wkhtmltopdf`. If you just write `pandoc -o myfile.pdf myfile.md`, Pandoc will do as if you had written `pandoc -o myfile.pdf -t latex --pdf-engine=pdflatex myfile.md`.

Of course, if you simply write `-o myfile.html` without `--pdf-engine`, Pandoc will simply generate an HTML file. Set `--pdf-engine` only if you want to make a PDF file from it.

**Reply by johnmapeson** [reply]
Thank you very much. This answer really helps me to understand how Pandoc works. Could I kindly ask you to copy and then paste it back so that the formatting will be fixed? Then I will delete my comment.

**Comment by badumont** [comment]
> Thank you very much. This answer is really very useful to better
> understand how Pandoc works.

You're welcome!

> Could I kindly ask you to copy and then
> paste it back so that the formatting will be fixed?

Sorry, but I receive all discussions in my mailbox and I am simply a bit tired of having to open my web browser and connect to Github just to be able to add rich text formatting to my message. If it seems important to you, I suggest that you write to Github's support and ask them to make Markdown markup be taken into account in replies sent by email ;-) In any case, thank you for having taken the time to do it yourself.
